services/gateway/gateway-service.mjs

```diff
@@ -245,53 +250,84 @@ class GatewayService {
    * - displayName: The name to show for this player.
    * - displayChar: The character asset to display.
    */
-  async onUserLocation(socket, msg) {
+  async onUserLocation(socket, origMsg) {
+    const msg = {};
     msg.playerID = socket.playerID;
     msg.displayName = socket.playerData.displayName;
     msg.displayChar = socket.playerData.displayChar;
+    msg.facing = origMsg.facing;
+    msg.mapCoord = MapCoord.fromObject(origMsg.mapCoord);
     // TODO: Check if facing is valid.
     // TODO: Check if movement is legal.
 
-    let mapSize = this.gameMap.getMapSize();
-    let lastMsg = { x:socket.playerData.x , y:socket.playerData.y}
-    if(lastMsg.x === undefined || lastMsg.y === undefined){
-      lastMsg.x = 0;
-      lastMsg.y = 0;
+    let lastCoord = socket.playerData.mapCoord;
+    let mapSize = this.gameMap.getMapSize(lastCoord.mapName);
+    if(lastCoord.x === undefined || lastCoord.y === undefined || lastCoord.mapName === undefined){
+      // Shouldn't happen, log an error.
+      console.error(`Invalid lastCoord in onUserLocation ${lastCoord}`);
+      return;
     }
     if(!this._movementRequestSpeedCheck(socket.playerData)){
+      console.warn(`Player ${msg.playerID} is speeding`);
+      return;
+    }
+    if (!(lastCoord.mapName === msg.mapCoord.mapName)) {
+      console.warn(`Player ${msg.playerID} changed map from ` +
+        `${lastCoord.mapName} to ${msg.mapCoord.mapName} without permission.`);
       return;
     }
     // target position is in the map
-    if(!this._borderCheck(msg,mapSize)){
-      // restrict user position
-      msg.x = lastMsg.x;
-      msg.y = lastMsg.y;
-      await this._broadcastUserLocation(msg);
+    if(!this._borderCheck(msg.mapCoord,mapSize)){
+      console.warn(`Player ${msg.playerID} is trying to go outside the map.`);
```

Comment:
Should we place the player right before the border instead of canceling their move?

Reply:
I thought the two is the same?

Reply:
Not exactly. They are the same only if the player's speed is less than 1 cell per tick.
If the player can move at, say, 2 cells per tick, and the distance between him and the border is 1 cell, he would not able to move if we cancel his move. However, I would expect that he can move 1 cell (that is, right in front of the wall).